pep-0688.rst

```diff
@@ -122,26 +122,40 @@ object, for example by the ``memoryview()`` constructor.
 It corresponds to the ``bf_getbuffer`` C slot.
 The Python signature for this method is
 ``def __buffer__(self, flags: int, /) -> memoryview: ...``. The method
-must return a ``memoryview`` object. If the method is called from C
-code, the interpreter extracts the underlying ``Py_buffer`` from the
-``memoryview`` and returns it to the C caller. Similarly, if the
-``__buffer__`` method is called on an instance of a C class that
+must return a ``memoryview`` object. If the ``bf_getbuffer`` slot
+is invoked on a Python class with a ``__buffer__`` method,
+the interpreter extracts the underlying ``Py_buffer`` from the
+``memoryview`` returned by the method
+and returns it to the C caller. Similarly, if Python code calls the
+``__buffer__`` method on an instance of a C class that
 implements ``bf_getbuffer``, the returned buffer is wrapped in a
 ``memoryview`` for consumption by Python code.
 
-The ``__release_buffer__`` method is called when a caller no
+The ``__release_buffer__`` method should be called when a caller no
 longer needs the buffer returned by ``__buffer__``. It corresponds to the
 ``bf_releasebuffer`` C slot. This is an
 optional part of the buffer protocol.
 The Python signature for this method is
 ``def __release_buffer__(self, buffer: memoryview, /) -> None: ...``.
 The buffer to be released is wrapped in a ``memoryview``. When this
 method is invoked through CPython's buffer API (for example, through
-``memoryview.release``), the passed ``memoryview`` is the same object
+calling ``memoryview.release`` on a ``memoryview`` returned by
+``__buffer__``), the passed ``memoryview`` is the same object
 as was returned by ``__buffer__``. It is
 also possible to call ``__release_buffer__`` on a C class that
 implements ``bf_releasebuffer``.
```

Comment:
This might be a different concern, but: what does calling ``__release_buffer__`` on such a class do?

Reply:
It calls `bf_releasebuffer`, if called with the same memoryview that `__buffer__` returned: https://github.com/JelleZijlstra/cpython/blob/pep688v2/Objects/typeobject.c#L7672. In fact it calls `memoryview.release()`, which takes care of ensuring that we don't double-call `bf_releasebuffer` (which would likely segfault).

Reply:
Would it be worth calling out in the PEP text that some calls to `.__release_buffer__()` will not result in a call to the underlying `bf_releasebuffer` slot? (e.g. if you call it twice, or call it with the wrong memoryview) I think the blanket statement that "The interpreter will ensure that such errors do not lead to memory leaks or memory safety violations" is sufficient, but happy to add more detailed wording.

Reply:
IMO it would be good to specify if these situations raise an exception, warn, or are silent no-ops.

Reply:
Done